aggregator/aggregator.go

```diff
@@ -157,24 +197,48 @@ func (a *Aggregator) AddOrCreate(key string, ts uint32, quantized uint, value fl
 
 // Flush finalizes and removes aggregations that are due
 func (a *Aggregator) Flush(ts uint) {
+	flushWaiting.Inc(1)
 	flushes.Add()
-	for k, proc := range a.aggregations {
-		if k.ts <= ts {
+	flushWaiting.Dec(1)
+	defer flushes.Done()
+
+	pos := -1 // will track the pos of the last ts position that was successfully processed
+	for i, quantized := range a.tsList {
+		if quantized > ts {
+			break
+		}
+		for key, proc := range a.aggregations[ts] {
 			results, ok := proc.Flush()
 			if ok {
 				if len(results) == 1 {
-					a.out <- []byte(fmt.Sprintf("%s %f %d", k.key, results[0].val, k.ts))
+					a.out <- []byte(fmt.Sprintf("%s %f %d", key, results[0].val, quantized))
+					a.numFlushed.Inc(1)
 				} else {
 					for _, result := range results {
-						a.out <- []byte(fmt.Sprintf("%s.%s %f %d", k.key, result.fcnName, result.val, k.ts))
+						a.out <- []byte(fmt.Sprintf("%s.%s %f %d", key, result.fcnName, result.val, quantized))
+						a.numFlushed.Inc(1)
 					}
 				}
 			}
-			delete(a.aggregations, k)
 		}
+		delete(a.aggregations, ts)
+		pos = i
+	}
+	// now we must delete all the timestamps from the ordered list
+	if pos == -1 {
+		// we didn't process anything, so no action needed
+		return
 	}
+	if pos == len(a.tsList)-1 {
+		// we went through all of them. can just reset the slice
+		a.tsList = a.tsList[:1]
```

Comment:
is this supposed to completely reset the slice? shouldn't it be `a.tsList = a.tsList[:0]`?

Reply:
oops